### PR TITLE
Support for modules partitions implementation unit

### PR DESCRIPTION
**Pull request by waruqi**
https://github.com/xmake-io/xmake/issues/4795

### CODE REVIEW

**Comment by waruqi** [comment]
### Msvc

```console
D:\projects\personal\xmake\tests\projects\c++\modules\partitions_implunit>xmake -rv
checking for C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe ... ok
checking for flags (cl_scan_dependencies) ... ok
checking for flags (cl_ifc_output) ... ok
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.cpp
checking for flags (-O2 -fp:fast) ... ok
[  0%]: <foo> generating.module.deps ModulesTutorial.cpp
[  0%]: <foo> generating.module.deps BasicPlane.Figures.ixx
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Point.ixx
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.ixx
checking for flags (-std:c++20) ... ok
checking for flags (-DNDEBUG) ... ok
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -scanDependencies build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.cpp.module.json BasicPlane.Figures-Rectangle.cpp -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.cpp.obj
BasicPlane.Figures-Rectangle.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -scanDependencies build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\ModulesTutorial.cpp.module.json ModulesTutorial.cpp -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e -Fobuild\.objs\foo\windows\x64\release\ModulesTutorial.cpp.obj
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -scanDependencies build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.ixx.module.json BasicPlane.Figures-Rectangle.ixx -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.ixx.obj
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -scanDependencies build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures.ixx.module.json BasicPlane.Figures.ixx -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures.ixx.obj
ModulesTutorial.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -scanDependencies build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ixx.module.json BasicPlane.Figures-Point.ixx -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Point.ixx.obj
BasicPlane.Figures-Rectangle.ixx
BasicPlane.Figures.ixx
BasicPlane.Figures-Point.ixx
[ 18%]: <foo> compiling.module.release BasicPlane.Figures:Point
checking for flags (cl_interface) ... ok
checking for flags (cl_internal_partition) ... ok
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\cl.exe" -c -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ifc -interface -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Point.ixx.obj BasicPlane.Figures-Point.ixx
checking for flags (cl_reference) ... ok
checking for flags (cl_header_unit_quote) ... ok
checking for flags (cl_header_unit_angle) ... ok
[ 27%]: <foo> compiling.module.release BasicPlane.Figures:Rectangle
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\cl.exe" -c -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -reference BasicPlane.Figures:Point=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ifc -TP -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.ifc -interface -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.ixx.obj BasicPlane.Figures-Rectangle.ixx
[ 36%]: <foo> compiling.module.release BasicPlane.Figures
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\cl.exe" -c -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -reference BasicPlane.Figures:Point=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ifc -reference BasicPlane.Figures:Rectangle=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.ifc -TP -ifcOutput build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures.ifc -interface -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures.ixx.obj BasicPlane.Figures.ixx
[ 63%]: compiling.release BasicPlane.Figures-Rectangle.cpp
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\cl.exe" -c -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -reference BasicPlane.Figures=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures.ifc -reference BasicPlane.Figures:Point=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ifc -reference BasicPlane.Figures:Rectangle=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.ifc -Fobuild\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.cpp.obj BasicPlane.Figures-Rectangle.cpp
[ 63%]: compiling.release ModulesTutorial.cpp
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\cl.exe" -c -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -reference BasicPlane.Figures=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures.ifc -reference BasicPlane.Figures:Point=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Point.ifc -reference BasicPlane.Figures:Rectangle=build\.gens\foo\windows\x64\release\rules\bmi\cache\modules\3319d13e\BasicPlane.Figures-Rectangle.ifc -Fobuild\.objs\foo\windows\x64\release\ModulesTutorial.cpp.obj ModulesTutorial.cpp
checking for flags (cl_sourceDependencies) ... ok
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\include\ostream(298): warning C4530: 使用了 C++ 异常处理程序，但未启用展开语义。请指定 /EHsc
checking for link.exe ... C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\link.exe
checking for the static library archiver (ar) ... link.exe
[ 81%]: archiving.release foo.lib
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Preview\\VC\\Tools\\MSVC\\14.40.33521\\bin\\HostX64\\x64\\link.exe" -lib -nologo -machine:x64 /opt:ref /opt:icf -out:build\windows\x64\release\foo.lib build\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.cpp.obj build\.objs\foo\windows\x64\release\ModulesTutorial.cpp.obj build\.objs\foo\windows\x64\release\BasicPlane.Figures-Point.ixx.obj build\.objs\foo\windows\x64\release\BasicPlane.Figures-Rectangle.ixx.obj build\.objs\foo\windows\x64\release\BasicPlane.Figures.ixx.obj
checking for cl.exe ... C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33521\bin\HostX64\x64\cl.exe
checking for the c compiler (cc) ... cl.exe
[100%]: build ok, spent 4.187s
```

**Comment by waruqi** [comment]
### Clang

But it still does not work for clang.

```console
ruki-2:partitions_implunit ruki$ xmake -rv
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.cpp
/usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Rectangle.cpp -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.cpp.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
[  0%]: <foo> generating.module.deps ModulesTutorial.cpp
/usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c ModulesTutorial.cpp -o build/.objs/foo/macosx/x86_64/release/ModulesTutorial.cpp.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
[  0%]: <foo> generating.module.deps BasicPlane.Figures.ixx
/usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.ixx
/usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Rectangle.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
[  0%]: <foo> generating.module.deps BasicPlane.Figures-Point.ixx
/usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Point.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Point.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
[ 18%]: <foo> compiling.module.release BasicPlane.Figures:Point
/usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Point.ixx.o BasicPlane.Figures-Point.ixx
[ 27%]: <foo> compiling.module.release BasicPlane.Figures:Rectangle
/usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.ixx.o BasicPlane.Figures-Rectangle.ixx
[ 36%]: <foo> compiling.module.release BasicPlane.Figures
/usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures.ixx.o BasicPlane.Figures.ixx
[ 63%]: compiling.release BasicPlane.Figures-Rectangle.cpp
/usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.cpp.o BasicPlane.Figures-Rectangle.cpp
[ 63%]: compiling.release ModulesTutorial.cpp
/usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -o build/.objs/foo/macosx/x86_64/release/ModulesTutorial.cpp.o ModulesTutorial.cpp
error: BasicPlane.Figures-Rectangle.cpp:5:16: error: unknown type name 'Rectangle'
    5 | int area(const Rectangle& r) { return width(r) * height(r); }
      |                ^
BasicPlane.Figures-Rectangle.cpp:6:18: error: unknown type name 'Rectangle'
    6 | int height(const Rectangle& r) { return r.ul.y - r.lr.y; }
      |                  ^
BasicPlane.Figures-Rectangle.cpp:7:17: error: unknown type name 'Rectangle'
    7 | int width(const Rectangle& r) { return r.lr.x - r.ul.x; }
      |                 ^
3 errors generated.
```

File: tests/projects/c++/modules/partitions_implunit/BasicPlane.Figures-Rectangle.cpp [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module;
+
+module BasicPlane.Figures:Rectangle;
```
Comment (dayu521) [inline tests/projects/c++/modules/partitions_implunit/BasicPlane.Figures-Rectangle.cpp:3]: - `module BasicPlane.Figures:Rectangle;` 改成`module BasicPlane.Figures:Rectangle.imp;`

- 增加导入 `import :Rectangle`

 三个平台是不是都可以编译成功?

**Comment by Arthapz** [comment]
> ### Clang/Gcc
> 
> 
> 
> But it still does not work for clang/gcc.
> 
> 
> 
> ```console
> 
> ruki-2:partitions_implunit ruki$ xmake -rv
> 
> [  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.cpp
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Rectangle.cpp -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.cpp.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
> 
> [  0%]: <foo> generating.module.deps ModulesTutorial.cpp
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c ModulesTutorial.cpp -o build/.objs/foo/macosx/x86_64/release/ModulesTutorial.cpp.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
> 
> [  0%]: <foo> generating.module.deps BasicPlane.Figures.ixx
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
> 
> [  0%]: <foo> generating.module.deps BasicPlane.Figures-Rectangle.ixx
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Rectangle.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
> 
> [  0%]: <foo> generating.module.deps BasicPlane.Figures-Point.ixx
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang-scan-deps --format=p1689 -- /usr/local/Cellar/llvm/17.0.6_1/bin/clang -x c++ -c BasicPlane.Figures-Point.ixx -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Point.ixx.o -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG
> 
> [ 18%]: <foo> compiling.module.release BasicPlane.Figures:Point
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Point.ixx.o BasicPlane.Figures-Point.ixx
> 
> [ 27%]: <foo> compiling.module.release BasicPlane.Figures:Rectangle
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.ixx.o BasicPlane.Figures-Rectangle.ixx
> 
> [ 36%]: <foo> compiling.module.release BasicPlane.Figures
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -x c++-module -fmodule-output=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures.ixx.o BasicPlane.Figures.ixx
> 
> [ 63%]: compiling.release BasicPlane.Figures-Rectangle.cpp
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -o build/.objs/foo/macosx/x86_64/release/BasicPlane.Figures-Rectangle.cpp.o BasicPlane.Figures-Rectangle.cpp
> 
> [ 63%]: compiling.release ModulesTutorial.cpp
> 
> /usr/local/Cellar/llvm/17.0.6_1/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -fno-delayed-template-parsing -DNDEBUG -fmodule-file=BasicPlane.Figures=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures.pcm -fmodule-file=BasicPlane.Figures:Point=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Point.pcm -fmodule-file=BasicPlane.Figures:Rectangle=build/.gens/foo/macosx/x86_64/release/rules/bmi/cache/modules/3319d13e/BasicPlane.Figures-Rectangle.pcm -o build/.objs/foo/macosx/x86_64/release/ModulesTutorial.cpp.o ModulesTutorial.cpp
> 
> error: BasicPlane.Figures-Rectangle.cpp:5:16: error: unknown type name 'Rectangle'
> 
>     5 | int area(const Rectangle& r) { return width(r) * height(r); }
> 
>       |                ^
> 
> BasicPlane.Figures-Rectangle.cpp:6:18: error: unknown type name 'Rectangle'
> 
>     6 | int height(const Rectangle& r) { return r.ul.y - r.lr.y; }
> 
>       |                  ^
> 
> BasicPlane.Figures-Rectangle.cpp:7:17: error: unknown type name 'Rectangle'
> 
>     7 | int width(const Rectangle& r) { return r.lr.x - r.ul.x; }
> 
>       |                 ^
> 
> 3 errors generated.
> 
> ```

I'll Check after work

File: tests/projects/c++/modules/partitions_implunit/BasicPlane.Figures-Rectangle.cpp (hunk 1 above)
Comment (waruqi) [reply tests/projects/c++/modules/partitions_implunit/BasicPlane.Figures-Rectangle.cpp:3]: 如果改成这样，clang 下即使没这个 patch ，当前版本也能编译成功，啥也不用改。
但是 gcc /msvc 编译不过

**Comment by waruqi** [comment]
> i don't think this can work, a partition is either an internal module partition unit or module interface partition unit see https://clang.llvm.org/docs/StandardCPlusPlusModules.html#module-and-module-unit msvc is wrong to allow that

The current patch is indeed a bit of a hack, because the dependency scanning of clang/msvc does not provide the corresponding dependency information. I don't know why msvc would allow this. If it is indeed non-standard, I will revert this patch.

**Comment by waruqi** [comment]
well, cmake does not work now. https://github.com/xmake-io/xmake/issues/4795#issuecomment-1993564239

**Pull request closed without merging**